test-harness/src/matcher/option_matcher.rs

```diff
@@ -0,0 +1,118 @@
+use std::fmt::Display;
+
+use crate::IntoMatcher;
+use crate::Matcher;
+
+use super::NeverMatcher;
+
+/// A matcher which may or may not contain a matcher.
+///
+/// If it does not contain a matcher, it never matches.
+pub struct OptionMatcher<M>(Option<M>);
```

Comment:
Could you instead `impl<M: Matcher> Matcher for Option<M>` or something like that?

Reply:
Not without writing my own `Display` replacement/wrapper trait at least:

```
error[E0277]: `std::option::Option<M>` doesn't implement `std::fmt::Display`
  --> test-harness/src/matcher/option_matcher.rs:13:30
   |
13 | impl<M: Matcher> Matcher for Option<M> {}
   |                              ^^^^^^^^^ `std::option::Option<M>` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `std::option::Option<M>`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
note: required by a bound in `Matcher`
  --> test-harness/src/matcher/mod.rs:30:20
   |
30 | pub trait Matcher: Display {
   |                    ^^^^^^^ required by this bound in `Matcher`
```